src/picongpu/include/particles/ionization/byCollision/ThomasFermi/ThomasFermi_Impl.hpp

```diff
@@ -166,13 +170,20 @@ namespace ionization
                 /* kernel call for weighted ion density calculation */
                 density->template computeValue< CORE + BORDER, DensitySolver >(*srcSpecies, currentStep);
                 dc.releaseData( SrcSpecies::FrameType::getName() );
+                EventTask densityEvent = density->asyncCommunication( __getTransactionEvent() );
+                densityEvent += density->asyncCommunicationGather( densityEvent );
```

Comment:
yes, this code had two bugs with regards to domain decomposition:
- the border was not filled correct (with additions from neighbor-GPU particles that have shape contributions) and
- since you access the density and energy density fields with respect to shape you also access the guard which was not created at all.

The first event, `asyncCommunication` builds a valid `BORDER`, the second event, `asyncCommunicationGather` builds a valid `GUARD`. Both APIs are also doxygen documented now since #2010